### PR TITLE
Cleanup MPICH TLS code

### PR DESCRIPTION
**Pull request by pavanbalaji**
## Pull Request Description

Cleanup MPICH TLS variables.

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [x] Add comments such that someone without knowledge of the code could understand
* [x] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
@pavanbalaji Does this PR prevent other PRs to use thread-local storage?

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/warnings

**Comment by pavanbalaji** [comment]
> @pavanbalaji Does this PR prevent other PRs to use thread-local storage?

It does, kind of.  MPL still supports TLS infrastructure.  Just the `MPID_` wrappers to those have been removed, and it also removes the macros that force a single TLS structure.  If we want to add other TLS variables, we need to add back the `MPID_` wrappers and simply declare the new variables as:

```
MPID_Thread_tls_t  myvar;
```

I can retain the `MPID_` wrappers, if folks prefer that.  But it seems unnecessary because we don't use that right now.

**Comment by hzhou** [comment]
> > @pavanbalaji Does this PR prevent other PRs to use thread-local storage?
> 
> It does, kind of. MPL still supports TLS infrastructure. Just the `MPID_` wrappers to those have been removed, and it also removes the macros that force a single TLS structure. If we want to add other TLS variables, we need to add back the `MPID_` wrappers and simply declare the new variables as:
> 
> ```
> MPID_Thread_tls_t  myvar;
> ```
> 
> I can retain the `MPID_` wrappers, if folks prefer that. But it seems unnecessary because we don't use that right now.

I believe @rzambre relies on the TLS. To keep it simple, one solution may be:
```
#ifdef MPL_TLS
MPL_TLS type myvar;
#else
/* either error or fallback to regular global, but probably error */
type myvar;
#endif
```

**Comment by pavanbalaji** [comment]
> I believe @rzambre relies on the TLS.

He does?  For what part?

> To keep it simple, one solution may be:
> 
> ```
> #ifdef MPL_TLS
> MPL_TLS type myvar;
> #else
> /* either error or fallback to regular global, but probably error */
> type myvar;
> #endif
> ```

We cannot assume all threads are MPL threads.  So, I assume you meant `MPID_` types instead of `MPL_` types.  But we can't simply do this as a declaration because not all threading models have compiler integration (e.g., argobots).

**Comment by hzhou** [comment]
> > I believe @rzambre relies on the TLS.
> 
> He does? For what part?

Per-thread request cache. And in general, it is convenient to have it for all kinds of experiments.

> We cannot assume all threads are MPL threads. So, I assume you meant MPID_ types instead of MPL_ types. But we can't simply do this as a declaration because not all threading models have compiler integration (e.g., argobots).

MPL wraps around threading interface, argobots included. It should leave MPL_TLS undefined if the compiler does not support it or in the case of argobots.

**Comment by pavanbalaji** [comment]
> Per-thread request cache. And in general, it is convenient to have it for all kinds of experiments.

Request caches are per VCI, not per thread.

> MPL wraps around threading interface, argobots included. It should leave MPL_TLS undefined if the compiler does not support it or in the case of argobots.

No, that's wrong.  If compiler support is not present, it'll not go into TLS-specific memory region, and it'll be incorrect.  For such cases, library calls must be used.

**Comment by pavanbalaji** [comment]
test:mpich/warnings
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
test:mpich/warnings
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
@hzhou Can you review this PR?

Given that there was so much debate as to whether we need TLS, I've brought back the TLS capabilities.  We can decide whether to remove it later as a separate PR, if needed.

**Comment by hzhou** [comment]
@pavanbalaji I liked your initial idea of letting `MPIR_Strerr` accept a caller-allocated buffer. Was that too messy to execute?

File: src/include/mpir_thread.h [hunk 1]
```diff
@@ -43,15 +43,15 @@ extern MPIR_Thread_info_t MPIR_ThreadInfo;
...
     char strerrbuf[MPIR_STRERROR_BUF_SIZE];
-} MPIR_Per_thread_t;
+} MPIR_Thread_tls_obj_t;
 
 #if defined(MPICH_IS_THREADED) && defined(MPL_TLS)
-extern MPL_TLS MPIR_Per_thread_t MPIR_Per_thread;
+extern MPL_TLS MPIR_Thread_tls_obj_t MPIR_Thread_tls_obj;
 #else
-extern MPIR_Per_thread_t MPIR_Per_thread;
+extern MPIR_Thread_tls_obj_t MPIR_Thread_tls_obj;
```
Comment (hzhou) [inline src/include/mpir_thread.h]: Are we concerned of the case when `MPICH_IS_THREADED` but no compiler support for TLS?

**Comment by pavanbalaji** [comment]
> @pavanbalaji I liked your initial idea of letting `MPIR_Strerr` accept a caller-allocated buffer. Was that too messy to execute?

No, not messy.  But if I remove that, there's nothing that we are putting in the TLS.  I can put a dummy variable in the TLS for now, if that's better.  FWIW, caller-allocated buffer is the direction POSIX is going in too, so that'd be consistent with the POSIX interface as well.

File: src/include/mpir_thread.h (hunk 1 above)
Comment (pavanbalaji) [reply src/include/mpir_thread.h]: Yes, not all threading libraries have compiler support, so we use the runtime interface for them.  But that code already exists in MPICH.  I'm not adding or removing it in this PR.

**Comment by hzhou** [comment]
> 
> 
> > @pavanbalaji I liked your initial idea of letting `MPIR_Strerr` accept a caller-allocated buffer. Was that too messy to execute?
> 
> No, not messy. But if I remove that, there's nothing that we are putting in the TLS. I can put a dummy variable in the TLS for now, if that's better. FWIW, caller-allocated buffer is the direction POSIX is going in too, so that'd be consistent with the POSIX interface as well.

I would suggest simply delete `MPIR_Thread_tls_obj` and simply leave the example as comments in, maybe, ` src/mpid/common/thread/mpidu_thread_fallback.h`. We can add specific tls key and obj as we add code that uses it.

**Comment by pavanbalaji** [comment]
> I would suggest simply delete `MPIR_Thread_tls_obj` and simply leave the example as comments in, maybe, ` src/mpid/common/thread/mpidu_thread_fallback.h`. We can add specific tls key and obj as we add code that uses it.

No, that'd remove a lot of necessary code (e.g., there's a hack function added to workaround a bug in the Intel compiler), which needs to be retained too.  I think the best approach is to leave a dummy variable in there.

**Comment by hzhou** [comment]
The hack function is a dummy function in the first place, so we are having a dummy function to fix a dummy variable. Don't you find it a bit funny? :). Besides, if we add new tls variable in a similar way, ie defining the variable separate from its usage, we will need one dummy function for each tls variable. How about just leave a comment on this bug (and fix) and reference to github issue?

**Comment by pavanbalaji** [comment]
> The hack function is a dummy function in the first place, so we are having a dummy function to fix a dummy variable. Don't you find it a bit funny? :). Besides, if we add new tls variable in a similar way, ie defining the variable separate from its usage, we will need one dummy function for each tls variable. How about just leave a comment on this bug (and fix) and reference to github issue?

If we are leaving a reference to a GitHub issue, why have the TLS infrastructure at all?  If someone needs it, they can retrieve it by reverting a specific patch.  They'll already see this GitHub issue when the PR gets merged in.

**Comment by hzhou** [comment]
yeah, it goes back to my original suggestion. When we consider using TLS, we really don't want to fallback to the tls_key, as it often defeats the purpose in the first place, right? So either we have compiler support or error

**Comment by pavanbalaji** [comment]
> yeah, it goes back to my original suggestion. When we consider using TLS, we really don't want to fallback to the tls_key, as it often defeats the purpose in the first place, right? So either we have compiler support or error

We cannot error out.  Other than POSIX threads, nothing else has compiler support.  For example, Argobots will not have compiler support.  So what you are suggesting is not a practical option.

I've pushed a new set of changes.  The MDTA code is separated into PR#4344.  In this version, there's a single patch that removes the TLS usage, so it's easy to bring it back if needed.

**Comment by pavanbalaji** [comment]
test:mpich/warnings
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/util/mpir_strerror.c [hunk 2]
```diff
@@ -14,24 +14,13 @@ int strerror_r(int errnum, char *strerrbuf, size_t buflen);
 #endif
 #endif
 
-/* provides a thread-safe version of strerror */
-const char *MPIR_Strerror(int errnum)
+/* wrapper for strerror_r to return the buffer used */
+const char *MPIR_Strerror(int errnum, char *strerrbuf)
```
Comment (hzhou) [inline src/util/mpir_strerror.c]: I am a bit concerned that we are assuming the buffer size by convention here. How about we further add the parameter of buffer size?

Comment (pavanbalaji) [reply src/util/mpir_strerror.c]: Yes, good point.  Such assumptions are bad coding practice.  I'll fix it.

**Comment by pavanbalaji** [comment]
test:mpich/warnings
test:jenkins/ch3/most
test:jenkins/ch4/most

**Pull request merged**